core/src/main/java/org/elasticsearch/action/support/replication/ReplicationTask.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.support.replication;
+
+import org.elasticsearch.common.inject.Provider;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.tasks.Task;
+
+import java.io.IOException;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Task that tracks replication actions.
+ */
+public class ReplicationTask extends Task {
+    private volatile String phase = "starting";
+
+    public ReplicationTask(long id, String type, String action, Provider<String> description, String parentNode, long parentId) {
+        super(id, type, action, description, parentNode, parentId);
+    }
+
+    /**
+     * Set the current phase of the task.
+     */
+    public void setPhase(String phase) {
+        this.phase = phase;
+    }
+
```

Comment:
I don't have a complete design here but I feel like phase should be called something more generic (like stage for example), moved all the way up to the task level and have a common expandable dictionary. The stage description is a bit tricky, but it looks like we have several stages like INITIALIZING, SENDING_CHILDREN_REQUESTS, WAITING_FOR_CHILDREN, CANCELLING, DONE, FAILED, etc that will be shared among many tasks. But then we will have some stages that will be important but very task specific. So, a few designs with pros and cons comes to mind. It can be an enum (to describe common stage) + string (to optionally describe a specific phase), or a class extending some common Stage class, or some ever-extending enum where tasks will add their specific stages (the simplest but not going to work for plugins), or maybe just a string and an utility class with a bunch of constants defining common phases.